### ISSUE TITLE
autodiff for user script functions aka torch.jit.script for autograd.Function

### ISSUE THREAD

**Issue by patrickpjiang** [pytorch/pytorch#22329]
I got an error when use jit.script on some new layers(implemented in c++):
```
RuntimeError:
attribute lookup is not defined on python value of type 'FunctionMeta':
@torch.jit.script_method
def forward(self, x):
    result = TopPoolFunction.apply(x)
             ~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    return result
```
here is my code:
```
class TopPoolFunction(Function):
    @staticmethod
    def forward(ctx, input):
        output = top_pool.forward(input)[0]
        ctx.save_for_backward(input)
        return output

    @staticmethod
    def backward(ctx, grad_output):
        input  = ctx.saved_variables[0]
        output = top_pool.backward(input, grad_output)[0]
        return output

class TopPool(torch.jit.ScriptModule):
    @torch.jit.script_method
    def forward(self, x):
        result = TopPoolFunction.apply(x)
        return result
```
What is the proper way to use script on new layers, HELPPPP!

**Comment by ngimel** [comment]
Unfortunately, custom autograd functions are not scriptable. A slightly hacky way of working around it is described here http://lernapparat.de/pytorch-traceable-differentiable/

**Comment by soumith** [comment]
cc: @suo worth fixing over the sprint

**Comment by ngimel** [comment]
I think the fix should be to generate the autograd tracking code that Tom had to write manually for the extensions with user-provided derivatives.yaml, if that's possible. That would help people writing extensions and using c++ APIs with extensions, and also greatly reduce the need for custom autograd functions - it looks like in 99% {?) of cases custom autograd functions are needed to make extensions differentiable.

**Comment by apaszke** [comment]
@t-vi has an implementation that implements scripting for `autograd.Function`s (and registers them with the source-to-source AD as well). We've discussed that a bit, so maybe we could use that instead?

**Comment by apaszke** [comment]
Also, I don't think that extending `derivatives.yaml` as an external API is a good idea. The existing `autograd.Function` syntax or the source-to-source nested-closure syntax are actually quite ok and are much less magical. It's fine to educate project developers about the codegen, but it's much harder to educate all the other users.

**Comment by t-vi** [comment]
So I'm planning to do a writeup this in one of the following weeks. If you look at [my branch](https://github.com/t-vi/pytorch/tree/autograd_script) you see that dealing with the `ctx` takes a lot of hacky rewriting (in compiler.cpp). Adam and I tentatively concluded that using a "return forward result + closure" approach similar to what's done in `symbolic_script.cpp` is better, roughly with the following reasoning:
- Contexts like `ctx` isn't that great a fit for the JIT - because each `ctx` would be it's own class, and the inplace modification are a bit awkward,
- rewriting (as the branch does) is a gross hack,
- making the user use tuples isn't great user experience.
- The closure-style approach does make the "capture for backward" all automatic for the user.

I was going to implement that, too, but if you sprint ahead...

**Comment by patrickpjiang** [comment]
> @t-vi has an implementation that implements scripting for `autograd.Function`s (and registers them with the source-to-source AD as well). We've discussed that a bit, so maybe we could use that instead?

Hi, could you pls post the link of that implementation, too many files to find.

**Comment by ngimel** [comment]
@apaszke, @t-vi thanks for the explanation. Indeed, symbolic_script like approach seems better.

**Comment by t-vi** [comment]
Just because @soumith discussed working on it during a sprint.

So I hope to code up something replacing the front some time next week, I'm currently planning to make it look like the symbolic_script:
```python
@torch.jit.autograd_script
def my_function(x):
    def backward(grad_out):
        return 99 * x * grad_out  # not the true gradient, mind you!
    return x**2, backward

@torch.jit.script
def test(x):
    res = my_function(x)
    return res
```
It's not 100% clean or  working yet, but it gets past the alias and shape analysis by importing from the other branch above, so now I have to inline the autograd scripted function and create its derivative, which should be very similar to the branch, too.

Relative to what autograd.Functions do, I don't have the SavedVariable variable checks, but (unless they hide somewhere in autodiff) I guess neither has the other autodiff code...

Best regards

Thomas

**Comment by ferrine** [comment]
Hi! Is there any workaround except C++?

**Comment by t-vi** [comment]
Reviving #22582, probably.

**Comment by ferrine** [comment]
Handling context should be tricky. I'd expect it all work with torch.jit.ScriptFunction subclassing rather than this decorator.

**Comment by t-vi** [comment]
I liked my solution very much. But you implement it, you get to decide.

**Comment by ferrine** [comment]
That was just what I was initially looking for. Any solution would be great

ср, 22 янв. 2020 г., 1:52 Thomas Viehmann <notifications@github.com>:

**Comment by t-vi** [comment]
Oh, incidentally, I did that, too, at first (could be in a branch somewhere but I forget), but it got messy quickly. The captures are much nicer to handle (and also nice from an API perspective).

**Comment by mechkb** [comment]
Any updates on this? 

I've a custom C++/ CUDA op that I'm trying to JIT using `jit.script`. I registered the `forward` and `backward` functions in the cpp file using `torch::RegisterOperators`. I'm able to successfully load these functions in python using `load_library()` and use them in my custom autograd op class. 

Now when I perform: 
```
@torch.jit.script
def op_func(x):
    # type: (Tensor) -> Tensor
    return AutoGradOpClass(x)
``` 

I get the following error: 
```
Tried to access nonexistent attribute or method 'saved_tensors',...
        dropout_prob_t          = ctx.saved_tensors
                                  ~~~~~~~~~~~~~~~~~ <--- HERE
```
Everything works if I remove the `torch.jit.script` decorator. Any help/ guidance/ update would be appreciated.

**Comment by n-gao** [comment]
Is there any update on this issue? It would be great to be able to implement our own functions and still use jit!

**Comment by suo** [comment]
No updates as of now unfortunately, but it's definitely on our radar!

**Comment by jorism1993** [comment]
Is there any update on this issue? It would be cool to use custom backward functies with jit!

**Comment by wyzero** [comment]
Is there any update about this issue?

**Comment by analyticsbot** [comment]
Hi - Is there any update about this issue?

**Comment by ngimel** [comment]
Python custom autograd functions are still not supported, however, it is possible to implement custom autograd function in c++ with `torch::autograd::Function` https://pytorch.org/tutorials/advanced/dispatcher.html#adding-autograd-support. See torchvision for examples of how this is done.

**Comment by nlgranger** [comment]
@ngimel can you please comment on why this is not a feature you plan to support? Is JIT developpment only targeting inference and not training?

**Comment by mmz1996** [comment]
Hi - Is there any update about this issue?

**Comment by supracharger** [comment]
Hi, I am also looking for any resolutions on the issue.

**Comment by ibmzzj** [comment]
As of PyTorch 2.2.0 custom backward in autograd.Function is still not supported on the Python side. This still has to be done in C++: as shown in https://pytorch.org/tutorials/advanced/dispatcher.html. Unfortunately in my case, after successfully implementing the custom operator with my custom backward function in C++, and enable TorchScript on the whole model, I only see a 1%~2% training time reduction from pure python implementation without TorchScript.

FWIW, I also encountered this 2019 doc http://lernapparat.de/pytorch-traceable-differentiable/ which seems to try to explain how this can be implemented in C++. The implementation looks different (and more complex) from the official dispatcher guide. In particular it uses TraceableFunction, which I cannot find any reference code, except a few old forked repo from torchfunc in github, which now no longer uses TraceableFunction either.

**Comment by ASFaust** [comment]
i have found a workaround for the case that you want to compute a custom gradient:

```
    f(x).detach() + x * g(x).detach() - (x * g(x)).detach()
```

where `f(x)` is your forward pass, and `g(x)` is your custom gradient. 

this lets you use torch jit script with custom gradients.

in the forward pass, it evaluates to `f(x)`, and in the backward pass, the gradient of this expression is `g(x)`.

it's not very memory or compute friendly, but at least you can run your model in jit script.
